### PR TITLE
Expose Obligations created during type inference.

### PR DESCRIPTION
**Pull request by gavinleroy**
This PR is a first pass at exposing the trait obligations generated and solved for during the type-check progress. Exposing these obligations allows for rustc plugins to use the public interface for proof trees (provided by the next gen trait solver).

The changes proposed track *all* obligations during the type-check process, this is desirable to not only look at the trees of failed obligations, but also those of successfully proved obligations. This feature is placed behind an unstable compiler option `track-trait-obligations` which should be used together with the `next-solver` option. I should note that the main interface is the function `inspect_typeck` made public in `rustc_hir_typeck/src/lib.rs` which allows the caller to provide a callback granting access to the `FnCtxt`.

r? @lcnr

### CODE REVIEW

File: compiler/rustc_infer/src/infer/mod.rs [hunk 1]
```diff
@@ -334,6 +334,8 @@ pub struct InferCtxt<'tcx> {
     pub intercrate: bool,
 
     next_trait_solver: bool,
+
+    pub fulfilled_obligations: Lrc<Lock<Vec<traits::FulfilledObligation<'tcx>>>>,
```
Comment (lcnr) [inline compiler/rustc_infer/src/infer/mod.rs:338]: Please change this to only be a field of the fulfillment context (and a method on the `Engine` trait to get them. Do not add it to the inference context. That will miss some obligations checked while probing, but given that we also sometimes prove by just using evaluate, this shouldn't change

However, having it on the infcx stores obligations from probes, which can leak inference variables, which is bad:tm: and potentially results in ICE

Comment (gavinleroy) [reply compiler/rustc_infer/src/infer/mod.rs:338]: Leaking inference variables is obviously not good. However, there is a motivation for catching obligations from within a probe. Take the following code:

```rust
#![feature(negative_impls)]

use rayon::prelude::*;
use std::sync::MutexGuard;

#[derive(Debug)]
struct NotSend;

impl !Send for NotSend {}
impl !Sync for NotSend {}

fn does_impl<T: IntoParallelIterator>() {}
fn check_impl() { 
    does_impl::<Vec<NotSend>>()
}

fn parallel_do(vs: Vec<NotSend>) {
    // Obligation `Vec<NotSend>: IntoParallelIterator` 
    // happens inside a probe. (At least IIUC)
    vs.into_par_iter();
}

fn main() {}
```

If my understanding of the trait solver is correct, when type-checking the expression `vs.into_par_iter()` the query `Vec<NotSend>: IntoParallelIterator` would happen inside of a probe. For the context of the "trait debugger" we're building, catching this obligation is important so that we can explain the call ambiguity to a user. 

Is there a way to get this obligation without storing the obligations on the `InferCtxt`?

**Comment by gavinleroy** [comment]
Following a suggestion on [Zulip](https://rust-lang.zulipchat.com/#narrow/stream/364551-t-types.2Ftrait-system-refactor/topic/exposing.20obligations/near/412552310), the inference context is cloned and stored alongside the captured obligations. However, I'm not sure I'm satisfied with this approach.

1. The `InferCtxt` must now implement `Clone`, which I see as undesirable, as nowhere else in the compiler should you clone the context.
2. The obligations still need to be stored in the `InferCtxt` as opposed to the `TyCtxt`. My understanding is because `rustc_middle` cannot depend on `rustc_infer`.
 
Here's an alternative proposal. Instead of using the callback on `inspect_typeck` to inspect the `FnCtxt`, we could give the the callback to the `InferCtxt`. When obligation tracking is enabled this callback is invoked with the `InferCtxt` on all root obligations in the `FulfillmentCtxt`. Thus making the callback of type `impl Fn(infcx: &InferCtxt<'tcx>, obligation: PredicateObligation<'tcx>)`. No cloning necessary, and external tools *should* be able to get all information necessary :tm:. I can of course test this claim with our tool locally.

@lcnr do you have thoughts on this approach vs what was discussed the other day?

**Comment by lcnr** [comment]
> 1. The `InferCtxt` must now implement `Clone`, which I see as undesirable, as nowhere else in the compiler should you clone the context.

https://doc.rust-lang.org/nightly/nightly-rustc/rustc_infer/infer/struct.InferCtxt.html#method.fork exists and should be used instead

> 2. The obligations still need to be stored in the `InferCtxt` as opposed to the `TyCtxt`. My understanding is because `rustc_middle` cannot depend on `rustc_infer`.

the type could be erased on the `TyCtxt`, however, I think your suggested alternative is to never store the goals anywhere but eagerly call the supplied closure? This seems preferable to me

File: compiler/rustc_trait_selection/src/solve/fulfill.rs [hunk 2]
```diff
@@ -32,11 +32,34 @@ pub struct FulfillmentCtxt<'tcx> {
...
+        result: &Result<(bool, Certainty, Vec<Goal<'tcx, ty::Predicate<'tcx>>>), NoSolution>,
+    ) {
+        if self.track_obligations {
+            if let Some(inspector) = infcx.obligation_inspector.get() {
+                let result = match result {
+                    Ok((_, c, _)) => Ok(*c),
+                    Err(NoSolution) => Err(NoSolution),
+                };
+                (inspector)(infcx, &obligation, result);
+            }
```
Comment (lcnr) [inline compiler/rustc_trait_selection/src/solve/fulfill.rs:61]: isn't the `Some(inspector)` check enough :thinking: why did you add a separate flag here?

Comment (gavinleroy) [reply compiler/rustc_trait_selection/src/solve/fulfill.rs:61]: In it's current state, it's technically possible to have a `Some` when the `track-trait-obligation` flag isn't set. I'll change it to check the flag before setting the option.

Comment (gavinleroy) [reply compiler/rustc_trait_selection/src/solve/fulfill.rs:61]: Ah, I see. The motivation to have the unstable flag was to make the changes seem less intrusive :) especially because using the inspector relies on using the next trait solver, which is behind a flag. But functionally there's no reason to have the flag.

Comment (gavinleroy) [reply compiler/rustc_trait_selection/src/solve/fulfill.rs]: I removed the unstable flag and added a comment to `inspect_typeck` which states that it only has an effect when using the next trait solver. That should be clear enough for its use. I'm not particularly worried that someone will rely on its behavior during normal compilation.

File: compiler/rustc_trait_selection/src/solve/fulfill.rs [hunk 3]
```diff
@@ -86,7 +102,9 @@ impl<'tcx> TraitEngine<'tcx> for FulfillmentCtxt<'tcx> {
                     root_obligation: obligation,
                 }
             })
-            .collect()
+            .collect();
+
+        errors
```
Comment (lcnr) [inline compiler/rustc_trait_selection/src/solve/fulfill.rs:107]: spurious change, please revert

File: compiler/rustc_hir_typeck/src/lib.rs [hunk 4]
```diff
@@ -149,14 +150,26 @@ fn diagnostic_only_typeck<'tcx>(tcx: TyCtxt<'tcx>, def_id: LocalDefId) -> &ty::T
         let span = tcx.hir().span(tcx.local_def_id_to_hir_id(def_id));
         Ty::new_error_with_message(tcx, span, "diagnostic only typeck table used")
     };
-    typeck_with_fallback(tcx, def_id, fallback)
+    typeck_with_fallback(tcx, def_id, fallback, None)
 }
 
-#[instrument(level = "debug", skip(tcx, fallback), ret)]
+/// Same as `typeck` but `inspect` is invoked on evaluation of each root obligation.
+/// Inspecting obligations only works with the new trait solver.
```
Comment (lcnr) [inline compiler/rustc_hir_typeck/src/lib.rs:157]: please also state that:

- only used by external tools
- not cached as it's not a query

**Review by lcnr — COMMENTED** [review]
r=me after nits
